Assets/BossRoom/Scripts/Client/UI/PopupPanel.cs

```diff
@@ -51,34 +60,61 @@ void ResetState()
         /// </summary>
         /// <param name="titleText">The title text at the top of the panel</param>
         /// <param name="mainText"> The text just under the title- the main body of text</param>
-        public static void ShowPopupPanel(string titleText, string mainText)
+        /// <param name="isCloseableByUser">If true, this popup can be closed by the user, else, it has to be closed manually</param>
+        /// <returns>The popup's id</returns>
+        public static long ShowPopupPanel(string titleText, string mainText, bool isCloseableByUser = true)
+        {
+            if (s_Instance == null)
+            {
+                Debug.LogError($"No PopupPanel instance found. Cannot display message: {titleText}: {mainText}.");
+                return -1;
+            }
+
+            return s_Instance.SetupPopupPanel(titleText, mainText, isCloseableByUser);
+        }
+
+        /// <summary>
+        /// Closes the currently displayed popup if it has the requested id.
+        /// </summary>
+        public static void RequestClosePopupPanel(long popupId)
         {
             if (s_Instance != null)
             {
-                s_Instance.SetupPopupPanel(titleText, mainText);
+                s_Instance.ClosePopupPanel(popupId);
             }
             else
             {
-                Debug.LogError($"No PopupPanel instance found. Cannot display message: {titleText}: {mainText}");
+                Debug.LogError($"No PopupPanel instance found. Cannot close popup.");
             }
         }
 
-        void SetupPopupPanel(string titleText, string mainText)
+        long SetupPopupPanel(string titleText, string mainText, bool isCloseableByUser = false)
         {
             if (m_IsPopupShown)
             {
                 Debug.Log("Trying to show popup, but another popup is already being shown.");
                 Debug.Log($"{titleText}. {mainText}");
+                return -1;
             }
-            else
-            {
-                ResetState();
 
-                m_TitleText.text = titleText;
-                m_MainText.text = mainText;
+            ResetState();
+
+            m_TitleText.text = titleText;
+            m_MainText.text = mainText;
+            m_ConfirmButton.SetActive(isCloseableByUser);
+            m_LoadingImage.SetActive(!isCloseableByUser);
 
-                gameObject.SetActive(true);
-                m_IsPopupShown = true;
+            gameObject.SetActive(true);
+            m_IsPopupShown = true;
+            m_DisplayedPopupId = s_NextPopupId++;
+            return m_DisplayedPopupId;
+        }
+
+        void ClosePopupPanel(long popupId)
+        {
+            if (m_DisplayedPopupId == popupId)
```

Comment:
This is not only in case of stacking, this is simply to make sure that we are closing the right popup. There are other cases than popup stacking where the current popup might be different than the one we want to close. I'll add the summary of the public static method to this one too to clarify